### ISSUE TITLE
"getKeypairFromEnvironment" ERROR

### ISSUE THREAD

**Issue by Joelvdb** [solana-developers/helpers#38]
Hey, 
i'm trying to use the function "getKeypairFromEnvironment" but i'm getting "Error: Invalid secret key in environment variable 'SECRET_KEY'!"

```
// 1. Generate a new keypair
const keypair1 = Keypair.generate();
await addKeypairToEnvFile(keypair1, "SECRET_KEY");

// 2. Load a keypair from a file
const keypair2 = await getKeypairFromEnvironment("SECRET_KEY");

```

The Error:
<img width="1097" alt="image" src="https://github.com/user-attachments/assets/8d2153fe-2da7-48e9-8568-b428579af1af">

**Comment by Blazing-Mike** [comment]
make sure you have the secret key in your .env file like this
`SECRET_KEY=[YOUR_SECRET_KEY]`

then import `"dotenv/config"` which enables you to access your env variables then the s so you can have something like this at

```js
import { getKeypairFromEnvironment } from "@solana-developers/helpers";
import "dotenv/config"
const keypair = getKeypairFromEnvironment("SECRET_KEY")

console.log(
  `✅ Finished! We've loaded our secret key securely, using an env file!`
);
```

**Comment by miawithcode** [comment]
I read the implement of `getKeypairFromEnvironment`

```ts
// @solana-developers/helpers/dist/index.js

var getKeypairFromEnvironment = (variableName) => {
  const secretKeyString = process.env[variableName];
  if (!secretKeyString) {
    throw new Error(`Please set '${variableName}' in environment.`);
  }
  let decodedSecretKey;
  try {
    decodedSecretKey = import_bs58.default.decode(secretKeyString);
    return import_web3.Keypair.fromSecretKey(decodedSecretKey);
  } catch (throwObject) {
    const error = throwObject;
    if (!error.message.includes("Non-base58 character")) {
      throw new Error(
        `Invalid secret key in environment variable '${variableName}'!`
      );
    }
  }
  try {
    decodedSecretKey = Uint8Array.from(JSON.parse(secretKeyString));
  } catch (error) {
    throw new Error(
      `Invalid secret key in environment variable '${variableName}'!`
    );
  }
  return import_web3.Keypair.fromSecretKey(decodedSecretKey);
};
```

I noticed these lines of code:
```ts
  if (!error.message.includes("Non-base58 character")) {
    throw new Error(
      `Invalid secret key in environment variable '${variableName}'!`
    );
  }
```

if the secret key is not Base58 encoded, then the code will throw error `Invalid secret key in environment variable`. Instead of throwing an error, we want to try the secret key with Uint8Array decode. So if we remove the exclamation mark `!` before `error.message.includes("Non-base58 character")` should fix the bug:

```ts
  if (error.message.includes("Non-base58 character")) {
    throw new Error(
      `Invalid secret key in environment variable '${variableName}'!`
    );
  }
```

You can find 'getKeypairFromEnvironment' in path  `node_modules/@solana-developers/helpers/dist/index.js:157`

**Comment by Joelvdb** [comment]
> I read the implement of `getKeypairFromEnvironment`
> 
> ```ts
> // @solana-developers/helpers/dist/index.js
> 
> var getKeypairFromEnvironment = (variableName) => {
>   const secretKeyString = process.env[variableName];
>   if (!secretKeyString) {
>     throw new Error(`Please set '${variableName}' in environment.`);
>   }
>   let decodedSecretKey;
>   try {
>     decodedSecretKey = import_bs58.default.decode(secretKeyString);
>     return import_web3.Keypair.fromSecretKey(decodedSecretKey);
>   } catch (throwObject) {
>     const error = throwObject;
>     if (!error.message.includes("Non-base58 character")) {
>       throw new Error(
>         `Invalid secret key in environment variable '${variableName}'!`
>       );
>     }
>   }
>   try {
>     decodedSecretKey = Uint8Array.from(JSON.parse(secretKeyString));
>   } catch (error) {
>     throw new Error(
>       `Invalid secret key in environment variable '${variableName}'!`
>     );
>   }
>   return import_web3.Keypair.fromSecretKey(decodedSecretKey);
> };
> ```
> 
> I noticed these lines of code:
> 
> ```ts
>   if (!error.message.includes("Non-base58 character")) {
>     throw new Error(
>       `Invalid secret key in environment variable '${variableName}'!`
>     );
>   }
> ```
> 
> if the secret key is not Base58 encoded, then the code will throw error `Invalid secret key in environment variable`. Instead of throwing an error, we want to try the secret key with Uint8Array decode. So if we remove the exclamation mark `!` before `error.message.includes("Non-base58 character")` should fix the bug:
> 
> ```ts
>   if (error.message.includes("Non-base58 character")) {
>     throw new Error(
>       `Invalid secret key in environment variable '${variableName}'!`
>     );
>   }
> ```
> 
> You can find 'getKeypairFromEnvironment' in path `node_modules/@solana-developers/helpers/dist/index.js:157`

yes, this worked for me thanks!

**Comment by rayjohnson1** [comment]
> > I read the implement of `getKeypairFromEnvironment`
> > ```ts
> > // @solana-developers/helpers/dist/index.js
> > 
> > var getKeypairFromEnvironment = (variableName) => {
> >   const secretKeyString = process.env[variableName];
> >   if (!secretKeyString) {
> >     throw new Error(`Please set '${variableName}' in environment.`);
> >   }
> >   let decodedSecretKey;
> >   try {
> >     decodedSecretKey = import_bs58.default.decode(secretKeyString);
> >     return import_web3.Keypair.fromSecretKey(decodedSecretKey);
> >   } catch (throwObject) {
> >     const error = throwObject;
> >     if (!error.message.includes("Non-base58 character")) {
> >       throw new Error(
> >         `Invalid secret key in environment variable '${variableName}'!`
> >       );
> >     }
> >   }
> >   try {
> >     decodedSecretKey = Uint8Array.from(JSON.parse(secretKeyString));
> >   } catch (error) {
> >     throw new Error(
> >       `Invalid secret key in environment variable '${variableName}'!`
> >     );
> >   }
> >   return import_web3.Keypair.fromSecretKey(decodedSecretKey);
> > };
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I noticed these lines of code:
> > ```ts
> >   if (!error.message.includes("Non-base58 character")) {
> >     throw new Error(
> >       `Invalid secret key in environment variable '${variableName}'!`
> >     );
> >   }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > if the secret key is not Base58 encoded, then the code will throw error `Invalid secret key in environment variable`. Instead of throwing an error, we want to try the secret key with Uint8Array decode. So if we remove the exclamation mark `!` before `error.message.includes("Non-base58 character")` should fix the bug:
> > ```ts
> >   if (error.message.includes("Non-base58 character")) {
> >     throw new Error(
> >       `Invalid secret key in environment variable '${variableName}'!`
> >     );
> >   }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > You can find 'getKeypairFromEnvironment' in path `node_modules/@solana-developers/helpers/dist/index.js:157`
> 
> yes, this worked for me. thanks! As i see the main problem is that i get error from "decodedSecretKey = import_bs58.default.decode(secretKeyString);" saying "import_bs58.default.decode" is not a function. might be a problem with my node version? I think removing the "!" just might be a small fix but might create other problems later.

The root problem here is that the package `import_bs58`, in the compiled javascript, is erroneously wrapping the contents of the `bs58` package inside of another `default { ... }` block, via passing `1` to the second parameter in the `__toESM` function.

```
var import_bs58 = __toESM(require("bs58"), 1);
```

If I pull the repository and mark the package.json as `"type": "commonjs"`, then build the code, the issue is fixed. But Im new to the whole `__toESM` method so not sure if this is the best thing to do here, but ill leave it as a breadcrumb for someone.

Note: I tried to push a PR but am receiving a 403, so I can't contribute it for further discussion on the PR.

**Comment by AlexMost** [comment]
Seems like published version of the last package contains old code version with `.mts` files. I have pulled this repo and run `npm run build` after comparison I have found that published version differs from the sources build in this way.

![diff](https://github.com/user-attachments/assets/479cc625-a658-40ef-a102-411879d68312)

Maybe simple rebuild and republishing package will solve the problem.

**Comment by miawithcode** [comment]
> Seems like the last published version of package contains an old code version with `.mts` files. I have pulled this repo and ran `npm run build`. After comparison I have found that version that is currently in npm package differs from the sources build in this way (the right side is a biuld from the latest master and it works ok):
> 
> ![diff](https://private-user-images.githubusercontent.com/1831767/355565687-479cc625-a658-40ef-a102-411879d68312.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjMxNzQyNjQsIm5iZiI6MTcyMzE3Mzk2NCwicGF0aCI6Ii8xODMxNzY3LzM1NTU2NTY4Ny00NzljYzYyNS1hNjU4LTQwZWYtYTEwMi00MTE4NzlkNjgzMTIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgwOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MDlUMDMyNjA0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTA0MzJkZGNmNDY1YmFiMDQ1MzIwYjg5MTgxMWVjNzVlODFiYmMxOGI5NWQxNjg4Y2RlMTEyMjkzZDQxZWU1OCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.GoWJ4qfRPFK_Q6ZsNSZ72vAnOuRP3b3aXS_q90ccWx4)
> 
> Maybe simple rebuild and republishing package will solve the problem.

Indeed! The issue has been solved. Thanks a lot!

**Comment by mbertogliati** [comment]
I have a different proposal, but I might be wrong.
The tsconfig.json is using the following configuration:
```
    "target": "es2022",
    "moduleResolution": "nodeNext",
    "module": "nodeNext",
```
That means that .ts files will be transpiled to .js files using ESModules (which uses `import`) instead of CommonJS (which uses `require`) and that makes sense since new software is moving out of CommonJS modules.

However, `tsup` is still compiling using CommonJS and that may be the cause of the problem.
I was able to fix it adding "--format esm" to the build script in package.json : `tsup src/index.ts --format esm --dts`  and then rebuild using `npm run build`. The resulting index.js is now using the `import` syntax, rising no errors.

This might break the code that still uses CommonJS modules.

**Comment by qubaitian** [comment]
@mikemaccana I am sorry to bother, but is anyone looking into this issue?

**Comment by Akash-Kumar-Sinha** [comment]
By running locally, this function is working properly 

```shell

export const getKeypairFromEnvironment = (variableName: string) => {
  const secretKeyString = process.env[variableName];
  if (!secretKeyString) {
    throw new Error(`Please set '${variableName}' in environment.`);
  }

  // Try the shorter base58 format first
  let decodedSecretKey: Uint8Array;
  try {
    decodedSecretKey = base58.decode(secretKeyString);
    return Keypair.fromSecretKey(decodedSecretKey);
  } catch (throwObject) {
    const error = throwObject as Error;
    if (!error.message.includes("Non-base58 character")) {
      throw new Error(
        `Invalid secret key in environment variable '${variableName}'!`
      );
    }
  }

  // Try the longer JSON format
  try {
    decodedSecretKey = Uint8Array.from(JSON.parse(secretKeyString));
  } catch (error) {
    throw new Error(
      `Invalid secret key in environment variable '${variableName}'!`
    );
  }
  return Keypair.fromSecretKey(decodedSecretKey);
};

const keypair = getKeypairFromEnvironment("SECRET_KEY");
```

But if i am importing it from `@solana-developers/helpers`  `import { getKeypairFromEnvironment } from "@solana-developers/helpers";`, it is throwing error of ```Error: Invalid secret key in environment variable 'SECRET_KEY'!```

![image](https://github.com/user-attachments/assets/0112be21-8e98-462a-8e16-5024f9c673d0)


This means that there is a problem with exporting the function from `@solana-developers/helpers`.

**Comment by mikemaccana** [comment]
Hi @qubaitian ! This should be fixed in 2.5.2 

https://github.com/solana-developers/helpers/commit/30233733de07a2f8005651918b3af1df25af4c50

**Comment by mikemaccana** [comment]
Can people confirm if this is still an issue on 2.5.2?

We had some build issues caused by `tsup` after 2.4

**Comment by mikemaccana** [comment]
Sorry! `tsup` started giving us troubles and moving to `tsc` had some hiccups. **This is fixed in 2.5.4.** Please raise a new issue if there are any more problems.
